kedro-org/kedro-viz
# Add 'preview' mechanism for SQL queries

## Description
If a data node relies on a file, then the user has the ability to see where that file is and have a preview into its data.

The same is not true for non-file datasets, like `SQLQueryDataset` or `APIDataset` who pull their data from non-file sources. In this instance, the user has a data node but it's unclear of its origin. 

This adds one in for those non-file sources. Essentially, a `filepath` attribute for datasets that have no files.

## TL:DR 
Resolves https://github.com/kedro-org/kedro-viz/issues/1521, adds the preview mechanism (already present for task nodes) to data nodes as well

## Development notes

Modifies DataNodeMetadata to have a `code` attribute. This is already present for TaskNodeMetadata, this is just making DataNode have the same capability.

This `code` attribute would store SQL command (whatever was used to generate data) or an address for API-sources. 

There are 8 non-file datasets available to users; this PR handles all 8 accordingly.

## QA notes

- All tests run normally.
- Added new test cases in `kedro_viz`-`test_models`-`test_utils.py` file to achieve 100% test coverage.
- To test your own data node's with 'preview', create a SQLQueryDataset-based data node. Exact reproduction is:
  - Create a new data node in `demo-project/conf/base/parameters/catalog_01_raw.yml` - make sure its type is SQLQueryDataset 

<p align="center">
<img width="340" alt="image" src="https://github.com/kedro-org/kedro-viz/assets/35583819/c2dba9b5-3987-4a83-a761-b7c868ff2135">
</p>

  - Add that node into `.../src/pipelines/data_ingestion/pipeline.py` 
  - If necessary, create a 'dummy' function and link it to the code in `pipeline.py` if kedro complains about missing functions
  - See [commit](https://github.com/kedro-org/kedro-viz/pull/1564/commits/8ca2d32b2ff1920b87fde01d9fb071f0a85727cc) for example codes to help you with all of the above

## What's next?

- Modifying Node's `filepath` attribute to contain a glimpse of the `code` attribute value (e.g. if dataset is API, the code would be the full source details whilst the `filepath` would be a simplistic version of it like 'www.api_source.com'). 
- Aiming to address this extension in a separate ticket though (PR is already large)

## Image Example
_Where the data node has some 'code' value associated with it, a switch is displayed_
&nbsp;
<img width="784" alt="image" src="https://github.com/kedro-org/kedro-viz/assets/35583819/132464e6-0b2f-4e70-b12a-2244ae7e1453">

_This displays a 'Source block' window; displays the underlying logic or source location that generated this data_
&nbsp;
<img width="953" alt="image" src="https://github.com/kedro-org/kedro-viz/assets/35583819/a697c347-b179-49ef-a6f2-6f6c38559f1b">

_This is in-line with the same mechanism used for task / function nodes_
&nbsp;
<img width="1025" alt="image" src="https://github.com/kedro-org/kedro-viz/assets/35583819/3401ddd6-0d63-4d0a-921c-11f20ff65da4">
